src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11251,14 +11251,45 @@ private bool ScanCast(bool forPattern = false)
                     // following a cast.
                     return !forPattern || this.CurrentToken.Kind switch
                     {
-                        SyntaxKind.PlusToken => true,
-                        SyntaxKind.MinusToken => true,
-                        SyntaxKind.AmpersandToken => true,
-                        SyntaxKind.AsteriskToken => true,
+                        SyntaxKind.PlusToken or
+                        SyntaxKind.MinusToken or
+                        SyntaxKind.AmpersandToken or
+                        SyntaxKind.AsteriskToken or
                         SyntaxKind.DotDotToken => true,
-                        _ => CanFollowCast(this.CurrentToken.Kind)
+                        SyntaxKind.IdentifierToken => !isBinaryPattern(),
+                        var tk => CanFollowCast(tk)
                     };
 
+                    bool isBinaryPattern()
```

Comment:
>isBinaryPattern [](start = 25, length = 15)

nit: local functions should be at the end of the method so as to not interrupt the flow of main logic.